File: sdk/runanywhere-android/src/main/java/com/runanywhere/sdk/models/ModelInfo.kt
package com.runanywhere.sdk.models

import java.io.File
import java.net.URL

/**
 * Information about a model
 */
data class ModelInfo(
    val id: String,
    val name: String,
    val format: ModelFormat,
    val downloadURL: URL? = null,
    var localPath: File? = null,
    val estimatedMemory: Long = 1_000_000_000, // 1GB default
    val contextLength: Int = 2048,
    val downloadSize: Long? = null,
    val checksum: String? = null,
    val compatibleFrameworks: List<LLMFramework> = emptyList(),
    val preferredFramework: LLMFramework? = null,
    val hardwareRequirements: List<HardwareRequirement> = emptyList(),
    val tokenizerFormat: TokenizerFormat? = null,
    val metadata: ModelInfoMetadata? = null,
    val alternativeDownloadURLs: List<URL>? = null,
    val additionalProperties: Map<String, Any> = emptyMap()
)

/**
 * Hardware requirement for a model
 */
data class HardwareRequirement(
    val type: HardwareType,
    val minimumSpecification: String,
    val recommendedSpecification: String? = null
)

/**
 * Hardware types
 */
enum class HardwareType {
    CPU,
    GPU,
    NPU,
    MEMORY,
    STORAGE
}

/**
 * Tokenizer format
 */
enum class TokenizerFormat {
    SENTENCEPIECE,
    BPE,
    WORDPIECE,
    UNKNOWN
}

/**
 * Model metadata
 */
data class ModelInfoMetadata(
    val version: String? = null,
    val description: String? = null,
    val author: String? = null,
    val license: String? = null,
    val tags: List<String> = emptyList(),
    val lastUsed: Long? = null,
    val downloadDate: Long? = null
) 

File: sdk/runanywhere-swift/Sources/RunAnywhere/Core/Models/ModelInfo.swift
import Foundation

/// Information about a model
public struct ModelInfo {
    public let id: String
    public let name: String
    public let format: ModelFormat
    public let downloadURL: URL?
    public var localPath: URL?
    public let estimatedMemory: Int64
    public let contextLength: Int
    public let downloadSize: Int64?
    public let checksum: String?
    public let compatibleFrameworks: [LLMFramework]
    public let preferredFramework: LLMFramework?
    public let hardwareRequirements: [HardwareRequirement]
    public let tokenizerFormat: TokenizerFormat?
    public let metadata: ModelInfoMetadata?
    public let alternativeDownloadURLs: [URL]?
    public var additionalProperties: [String: Any] = [:]

    public init(
        id: String,
        name: String,
        format: ModelFormat,
        downloadURL: URL? = nil,
        localPath: URL? = nil,
        estimatedMemory: Int64 = 1_000_000_000, // 1GB default
        contextLength: Int = 2048,
        downloadSize: Int64? = nil,
        checksum: String? = nil,
        compatibleFrameworks: [LLMFramework] = [],
        preferredFramework: LLMFramework? = nil,
        hardwareRequirements: [HardwareRequirement] = [],
        tokenizerFormat: TokenizerFormat? = nil,
        metadata: ModelInfoMetadata? = nil,
        alternativeDownloadURLs: [URL]? = nil,
        additionalProperties: [String: Any] = [:]
    ) {
        self.id = id
        self.name = name
        self.format = format
        self.downloadURL = downloadURL
        self.localPath = localPath
        self.estimatedMemory = estimatedMemory
        self.contextLength = contextLength
        self.downloadSize = downloadSize
        self.checksum = checksum
        self.compatibleFrameworks = compatibleFrameworks
        self.preferredFramework = preferredFramework
        self.hardwareRequirements = hardwareRequirements
        self.tokenizerFormat = tokenizerFormat
        self.metadata = metadata
        self.alternativeDownloadURLs = alternativeDownloadURLs
        self.additionalProperties = additionalProperties
    }
}
